Media CdnBaseUrl Configuration

deanmarcussen opened the pull request:
This is probably not ready, but it works!

Have implemented a CdnUrl appSetting which is prepended to Local files only (not blob storage) when served.

Also adjusted tenant appSettings to load based on environment as well so it can be disabled / enabled at a tenant level in development.

I put it live to make sure it worked in production and you can see it working at https://www.theappspajamas.com/docs/getting-started/upload-assets
media is served from the cdn on https://media.theappspajamas.com

Here's some screen shots  

First is when is-cache folder has been cleared and cdn has never resolved the file.
It's slow, but the cdn doesn't timeout which is what I wanted to be sure of
![resize-on-fly](https://user-images.githubusercontent.com/13782679/56090645-cd131600-5e9c-11e9-8e26-20f6d93e2468.png)


Second is reload while clearing local browser cache, after media has been resolved by IS resizer, and cdn has in its cache. Much faster!
![from-cdn](https://user-images.githubusercontent.com/13782679/56090651-d603e780-5e9c-11e9-9b5d-dae0e9c8952d.png)

Also noticed that ImageSharp moves even non resized images to its is-cache folder. Surprised me. 
Does mean it has the potential to grow quite large, which isn't necessary if the cdn has fronted it.

Will wait for #3200 to come in and maybe can look at getting working with blob storage on another PR.

jtkech commented on src/OrchardCore/OrchardCore/Environment/Shell/Configuration/ShellConfigurationSources.cs:
```diff
@@ -10,18 +10,21 @@ namespace OrchardCore.Environment.Shell.Configuration
+                .AddJsonFile(Path.Combine(_container, tenant, "appsettings.json"), optional: true)
+                .AddJsonFile(Path.Combine(_container, tenant, $"appsettings.{_environment}.json"), optional: true);
         }
 
```
Hmm, not sure because `appsettings.json` in the tenant folder is mutable, when setup / edit a tenant this is where we save **config values** that are not pre-configured (from other immutable sources) or different. Idem for **setting values** but to the shared `tenants.json`. (I did a comment on gitter related to this).

So here, in a given env you may get a diferent value, then if you setup / edit and save the tenant this value will be saved in `appsettings.json`, then this value may be used when running in another env.

deanmarcussen replied:
ah, ok. I can revert that. 

If a configuration is stored in tenant appsettings, how can an environment setting override at tenant level? I will see if I can find the gitter comment and understand.

jtkech replied:
In another config source, e.g in an `appsettings.{env}.json` at the root level or just under the `App_Data` folder. This just under the `OrchardCore` section to apply to all tenants, or under a subsection (by using the tenant name) to apply to this tenant. Note: to be recognized as a tenant subsection, you have to define a state value (used as a marker) e.g `"State": Uninitialized`.

deanmarcussen replied:
@jtkech Yes, I revert that then. However will wait until you have chance to review #3581 as it will break this - will need more virtual path mangling!

So **not ready**

deanmarcussen commented:
I've reverted the changes to `ShellConfigurationSources`, and refactored this to work with the asp-append-version support. 

It works well using `https://localhost:44300` as the `CdnBaseUrl`, but I'll pull it into a production site and test asap, to be sure.

I've used the name `CdnBaseUrl` to keep the naming the same as what I did for Resources, in the `ISite.CdnBaseUrl` property, but they remain separate properties (Media through `IShellConfiguration` and Resources through `ISite`), to allow different cdn's for Media and Static Assets

jtkech commented:
As i remember so not sure but maybe you just need to use the `CdnBaseUrl` in `MapPathToPublicUrl()` this to generate the rendered url. Then, when a given cdn point to OC and does a request, i think that `MapPublicUrlToPath` would have to work as before.

Maybe also it is not necessary to change `ShellFileVersionProvider` but i understand the problem in e.g the `ImageTagHelper`. So, maybe don't change also `MapPathToPublicUrl()` so that we could use the version provider as before. And maybe another method `UseCdn(string publicUrl)` or just a property `CdnBaseUrl` used by the image tag helper / filter to get / build the full url.

Note: Maybe a `MediaOptions` config class that could be injected and used independently in different places, and at the level they need to be used.

For blob, we already use a full uri for the `publicUrlBase`, so need to be updated. As i remember, we could re-use the same media services implementations, only the `BlobFileStore` would have to do the bridge, and if we want to point directly to the blob we could use the new `CdnBaseUrl` config value.

deanmarcussen commented:
Thanks @jtkech 

So yes, `MapPublichPathToPath` will work as before, if preferred. It is actually only used in one place, which doesn't need any change for it still to work. The reason I update it to consider both possible incoming strings, was merely for extensibility, as I don't know what other uses people will choose to put it too.

Re: `ShellFileVersionProvider`
I consider two options when writing this

1. Use a method `UseCdn()` in `IMediaFileProvider` or
2. Make `ShellFileVersionProvider` responsible for mangling the path

I went with number 2 because
* Experience implementing CDN mapping has been that if you provide two methods to do the mapping someone will always forget to use the second method, and the CDN will not be used
* I wanted to also be able to support the `<img asp-append-version="true"/>` mvc tag helper, and using the `OrchardHelper.AssetSrc` would not be able to make use of a `UseCdn()` method 

`MediaOptions` I think this is a great idea, as we use those configuration options in many places now, and it makes more sense to register them once (and allow registration changes) than only provide via `IShellConfiguration`. I think it may need a small extension method for it's registration to set the defaults, but I am happy to put in as part of this PR.

**Note** For blob I do nothing in this PR as still waiting for design decisions from @sebastienros Right now it will continue to use `publicUrlBase` and does not make use of the `CdnBaseUrl` property at all. However when we do the work on `BlobFileStore` maybe we replace the `PublicHost` property and can share `CdnBaseUrl` property - depends on the design decision. So right now `BlobStorage` continues to work as before

deanmarcussen commented:
Actually I think I may be wrong about the assetsrc helper as we implement asp-append-version as a feature of it. Still must look and check it when I am back in a couple days

jtkech commented:
Okay i understand, but i still feel that `ShellFileVersionProvider` is not the right place to use `CdnBaseUrl`. I will think about it this week-end, and maybe also about the blob part.

deanmarcussen commented:
@jtkech So I have a better look and yes, I can see why `ShellFileVersionProvider` is not necessarily the best place to do this.

To do as you suggest with another method like `MapToCdnUrl` in the `IMediaFileStore` is the more logical approach, and generally better, as it would just provide the correct url at the correct time etc.

There is only one (maybe two w/blob see below) use cases for doing something with `ShellFileVersionProvider` to allow `ShellFileVersionProvider` to work in conjunction with the Mvc `ImageTagHelper`, but maybe this is ok not to support? 

e.g. this will work
    `<img src="@Orchard.AssetUrl(assetPath: "/portfolio/02-full.jpg", appendVersion: true)"/>`
but this will not
    `<img src="@Orchard.AssetUrl("/portfolio/02-full.jpg")" asp-append-version="true" />`

So, I see what you think about the need to support the Mvc `ImageTagHelper`. If we want to support it still, we could decouple the `CdnBaseUrl` from `ShellFileVersionProvider` with a `CdnPathProvider`, perhaps.

For blob things remain uncertain, as it does not implement a `IStaticFileProvider` at all, and probably should not. But to support `asp-append-version` with it, which I think there is a use case for (hopefully via blob properties with the MD5 hash it already calculates), requires further working on the `ShellFileVersionProvider`. We have both said a few times, **blob later**, and I have said it again already on this PR. Perhaps it is better if it is **blob time next**, and this PR should either include it, or this PR should wait until we have worked on blob?

jtkech commented:
> but this will not work
<img src="@Orchard.AssetUrl("/portfolio/02-full.jpg")" asp-append-version="true" />

Oh yes, so maybe i was wrong. I worked on other things but i will think about it soon.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStore.cs:
```diff
@@ -79,17 +81,29 @@ public Task CreateFileFromStream(string path, Stream inputStream, bool overwrite
+            if (publicUrl.StartsWith(_cdnBaseUrl))
+            {
+                var resolvedPath = publicUrl.Substring(_cdnBaseUrl.Length);
+                if (resolvedPath.StartsWith(_publicUrlBase))
```
culture ordinal ignore case

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStore.cs:
```diff
@@ -79,17 +81,29 @@ public Task CreateFileFromStream(string path, Stream inputStream, bool overwrite
         public string MapPathToPublicUrl(string path)
         {
-            return _publicUrlBase.TrimEnd('/') + "/" + this.NormalizePath(path);
+            return _cdnBaseUrl + _publicUrlBase.TrimEnd('/') + "/" + this.NormalizePath(path);
```
Where do we ensure that it has a trailing `/` ?

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Modules/FileProviders/ICdnBaseUrlProvider.cs:
```diff
@@ -0,0 +1,7 @@
+namespace OrchardCore.Abstractions.Modules.FileProviders
+{
+    public interface ICdnBaseUrlProvider
```
Why is it necessary?

sebastienros commented:
Did we already merge the CdnBaseUrl support for resources ?

deanmarcussen commented:
Yes, we did merge `CdnBaseUrl` for resources, but this PR is still **not ready**

When you get through all those gitter messages 😄 you'll find a short novel from me about the design for blob resizing, and some research on it.

The short version is this PR would probably need refactoring to support blob, so I've taken a branch from here, and have a working implementation of blob w/resizing, cache busting, and cdn support, which is just waiting for some design considerations, before I tidy it up and get it ready for a PR review.

deanmarcussen commented:
Moved to #4025
